src/api/routes/v1/get-network.ts

```diff
@@ -0,0 +1,237 @@
+}
+
+/**
+ * Checks whether the UNL of the node is recorded in the networks table. If so,
+ * then it returns the network associated with the UNL.
+ *
+ * @param unl - The UNL of the node.
+ * @returns Whether the UNL of the node has been seen before.
+ */
+async function getNetworkFromUNL(unl: string): Promise<string | undefined> {
+  const result = await query('networks')
+    .select('id')
+    .where('unls', 'like', `%${unl}%`)
+  return result.length > 0 ? result[0].id : undefined
+}
+
+/**
+ * Checks whether the public key of the node is recorded in the crawls table. If
+ * so, then it returns the network associated with the public key.
+ *
+ * @param publicKey - The public key of the node.
+ * @returns The network associated with the public key of the node. Undefined
+ * if the public key has not been seen by the network.
+ */
+async function getNetworkFromPublicKey(
+  publicKey: string,
+): Promise<string | undefined> {
+  const result = await query('crawls')
+    .select('public_key', 'networks')
+    .where('public_key', '=', publicKey)
+  return result.length > 0 ? result[0].networks : undefined
+}
+
+/**
+ * Adds the node (and its corresponding network) to the networks table.
+ *
+ * @param url - The URL endpoint of the node.
+ * @param unl - The UNL of the node.
+ * @returns The ID of the new network.
+ */
+async function addNode(url: string, unl: string | null): Promise<string> {
+  const newNetwork = (Number(maxNetwork) + 1).toString()
+  maxNetwork += 1
+
+  const network: Network = {
+    id: newNetwork,
+    entry: url,
+    port: 51235,
+    unls: unl ? [unl] : [],
+  }
+  await query('networks').insert({
+    ...network,
+    unls: network.unls.join(','),
+  })
+
+  const crawler = new Crawler()
+  void crawler.crawl(network)
+
+  return newNetwork
+}
+
+/**
+ * Fetch network ID from network entry. If the network doesn't exist, adds it with a new ID.
+ *
+ * @param req - Express request.
+ * @param res - Express response.
+ * @returns The Express response.
+ */
+export default async function getNetwork(
+  req: Request,
+  res: Response,
+): Promise<Response> {
+  try {
+    const { entryUrl } = req.params
+
+    // fetch crawl
+    const crawl = await fetchCrawls(entryUrl)
+
+    // check UNL
+    const { node_unl } = crawl
+    if (node_unl != null) {
+      const unlNetwork = await getNetworkFromUNL(node_unl)
+      // eslint-disable-next-line max-depth -- Necessary here
+      if (unlNetwork != null) {
+        return res.send({
+          result: 'success',
+          network: unlNetwork,
+        })
+      }
+    }
+
+    // check if node public key is already recorded
+    const { public_key } = crawl.this_node
+    const publicKeyNetwork = await getNetworkFromPublicKey(public_key)
+    if (publicKeyNetwork != null) {
+      return res.send({
+        result: 'success',
+        network: publicKeyNetwork,
+      })
+    }
+    // add node to networks list
+    const newNetwork = await addNode(entryUrl, node_unl)
```

Comment:
Can we add a check that the crawl was valid (i.e. produced > `n`) nodes. Otherwise I think I could just spam random entries that will be saved in `addNode`. 

Also, I think `node_unl` can be undefined here (if node_unl == null). Should that be a possibility, or can we save an IP w/o a node. 

We should ensure that we got a valid crawl, a valid UNL, and a valid public_key before we save a node, else anyone can write arbitrary information to the database.

Reply:
I would also err on the side of making it more difficult to add your chain to the VHS database, over making it too easy.

Reply:
I don't quite understand how you could spam random entries - the `getNetworkFromPublicKey` call checks whether the node already exists in the known networks.